### DISCUSSION TITLE
Batching resolvers (data fetchers)

### DISCUSSION THREAD

**Discussion by jugalthakkar** [graphql-java-kickstart/graphql-java-tools#389] [Q&A] [answered]
I really like the way that graphql-java-tools minimised boilerplate code and helped in getting started very quickly. 
I am at a stage where I want to optimise my code for performance and came across batching in graphql-java. This seems to be what I am looking for, but unable to find example to do this in graphql-java-tools. 

Is there a documented way of achieving this? Or an example of how we can mix and use both graphql-java-tools and graphql-java together as per requirement?

**Comment by oliemansm** [comment]
Are you using just graphql-java-tools or are you using it in combination with graphql-spring-boot-starter? There's an example on how to use data loaders in that repo, see: https://github.com/graphql-java-kickstart/graphql-spring-boot/tree/master/example-request-scoped-dataloader.

**Comment by vojtapol** [comment]
What you are talking about is not really batching (the batch strategy is deprecated) but [data loading](https://github.com/graphql-java/java-dataloader). That concept is unrelated to this library. Data loaders are usually accessible via `DataFetchingEnvironment` which is a `graphql-java` class and you can access an instance of it in each of your `graphql-java-tools` resolvers.

**Comment by benmccann** [comment, marked as answer]
I had the same question while trying to get this setup with Micronaut. When you read the GraphQL Java docs they have a whole [chapter on "batching"](https://www.graphql-java.com/documentation/v14/batching/) and make it sound very important that you implement a `BatchLoader` for each type. When using graphql-java-tools it looks like we use `GraphQLResolver` instead of `DataFetcher` and so there's this big open question in the docs of do we use `BatchLoader` or does graphql-java-tools have its own version of that too? If we do use `BatchLoader` how to we integrate it since some of the APIs are different (e.g. now we use `SchemaParser`, `SchemaParserBuilder`, etc. instead of the graphql-java versions)

It looks like several others have had this question as well: https://github.com/graphql-java-kickstart/graphql-java-tools/issues/370 https://github.com/graphql-java-kickstart/graphql-java-tools/issues/291 https://github.com/graphql-java-kickstart/graphql-java-tools/issues/290 https://github.com/graphql-java-kickstart/graphql-java-tools/issues/151 https://github.com/graphql-java-kickstart/graphql-java-tools/issues/58. There's a lot of different parts to put together a working GraphQL server and it's a bit tricky to figure out what to use from each library. Perhaps we could add something to the docs?

It looks like https://github.com/graphql-java-kickstart/graphql-java-tools/issues/58 has the best discussion of how this works

**Comment by benmccann** [comment]
I just got this working after much struggling. The key point I was missing is that your `GraphQLResolver` has to explicitly call the `DataLoader`. This wasn't very clear to me. Most of the examples that are available are for Spring and don't use `graphql-java-tools`, so don't have any examples that use a `GraphQLResolver`. It's also a little confusing because eventhough it's a `BatchLoader`, you're only calling it with a single key

**Comment by benmccann** [comment]
Now that I know how things work, https://github.com/graphql-java-kickstart/samples/tree/master/tools-hello-world doesn't seem very realistic. Everytime you get a character from the repository all their friends are returned too. But that's not really how you'd actually set things up if your app is backed by an RDBMS, which is the common case, because it'd be wasteful to fetch the friends if the GraphQL query doesn't request them. Instead, I think this would be the perfect place to demonstrate the usage of a `BatchLoader`.

**Comment by vojtapol** [comment]
Glad you figured it out. I took me a while too as you can see in #58.

The example you linked is correct. It might not be very realistic but it keeps things simple. Maybe we should add a separate example that demonstrates the use of data loader. There will always be a little bit of overfetching or underfetching when you back the GraphQL API with an RDBMS.

**Comment by mxmlnglt** [comment]
> Now that I know how things work, https://github.com/graphql-java-kickstart/samples/tree/master/tools-hello-world doesn't seem very realistic. Everytime you get a character from the repository all their friends are returned too. But that's not really how you'd actually set things up if your app is backed by an RDBMS, which is the common case, because it'd be wasteful to fetch the friends if the GraphQL query doesn't request them.

Hello @benmccann, regarding the example you're referring to: what do you mean by *all their friends are returned too*?

Is that what's done in the methods in the `QueryResolver`?
https://github.com/graphql-java-kickstart/samples/blob/master/tools-hello-world/src/main/java/hello/resolvers/Query.java

e.g. for a single search (by id), retrieving the whole data?
```
  public Character character(String id) {
    return characterRepository.getCharacters().get(id);
  }
```

**Comment by benmccann** [comment]
See here: https://github.com/graphql-java-kickstart/samples/blob/5a75ad6e025b06d35abf9a66d558fef3d81c7df9/tools-hello-world/src/main/java/hello/CharacterRepository.java#L37
